File: 1578-minimum-time-to-make-rope-colorful/1578-minimum-time-to-make-rope-colorful.py
class Solution:
    def minCost(self, colors: str, neededTime: List[int]) -> int:
        result = 0
        index = 0
        current = []
        
        while index < len(neededTime):
            if len(current) != 0 and colors[index] != current[-1][1]:
                current = sorted(current, key=lambda x: -x[0])
                while len(current) > 1:
                    item = current.pop()
                    result += item[0]
                
                current = []
            
            current.append((neededTime[index], colors[index]))
            index += 1
        
        current = sorted(current, key=lambda x: -x[0])
        while len(current) > 1:
            item = current.pop()
            result += item[0]
        
        return result

File: 1578-minimum-time-to-make-rope-colorful/1578-minimum-time-to-make-rope-colorful.swift
class Solution {
    typealias Balloon = (Int, String)
    
    func minCost(_ colors: String, _ neededTime: [Int]) -> Int {
        let hues = colors.map { String($0) }
        
        var result = 0
        var i = 0
        var current: [Balloon] = []
        
        while i < neededTime.count {
            if current.count != 0 && hues[i] != current[current.count-1].1 {
                current.sort { $0.0 > $1.0 }
                while current.count > 1 {
                    let item = current.removeLast()
                    result += item.0
                }
                
                current.removeAll()
            }
            
            current.append((neededTime[i], hues[i]))
            i += 1
        }
        
        current.sort { $0.0 > $1.0 }
        while current.count > 1 {
            let item = current.removeLast()
            result += item.0
        }
        
        return result
    }
}